### PR TITLE
Allow handling of custom http verbs. Add hookAnyCustom and hookRouteCustom

### PR DESCRIPTION
**Pull request by merc1031**
Address #61 by allowing custom http verbs for actions.

### CODE REVIEW

File: src/Web/Spock/Internal/Wire.hs [hunk 1]
```diff
@@ -54,6 +54,17 @@ import qualified Network.Wai.Parse as P
 instance Hashable StdMethod where
     hashWithSalt = hashUsing fromEnum
 
+data SpockMethod
```
Comment (agrafix) [inline src/Web/Spock/Internal/Wire.hs]: I think prefixing the constructors would be good: MethodStandard and MethodCustom e.g.?

File: src/Web/Spock/Internal/Wire.hs [hunk 2]
```diff
@@ -54,6 +54,17 @@ import qualified Network.Wai.Parse as P
...
+   = Standard StdMethod
+   | Custom T.Text
+     deriving Eq
+
+instance Hashable SpockMethod where
```
Comment (agrafix) [inline src/Web/Spock/Internal/Wire.hs]: This should be generated automatically with generics

File: src/Web/Spock/Internal/Wire.hs [hunk 3]
```diff
@@ -401,11 +412,16 @@ buildMiddleware mLimit registryIf registryLift spockActions =
...
-                  do let allActions = getMatchingRoutes stdMethod (Wai.pathInfo req)
-                     runResourceT $ withInternalState $ \st ->
-                         handleRequest stdMethod mLimit registryLift allActions st coreApp req respond
+            parseSpockMethod (Wai.requestMethod req) $ \method -> do
+                let allActions = getMatchingRoutes method (Wai.pathInfo req)
+                runResourceT $ withInternalState $ \st ->
+                    handleRequest method mLimit registryLift allActions st coreApp req respond
        return $ spockMiddleware . app
+
+parseSpockMethod :: forall t. Method -> (SpockMethod -> t) -> t
```
Comment (agrafix) [inline src/Web/Spock/Internal/Wire.hs]: rename to withSpockMethod

File: src/Web/Spock/Internal/Wire.hs [hunk 4]
```diff
@@ -401,11 +412,16 @@ buildMiddleware mLimit registryIf registryLift spockActions =
...
            app :: Wai.Application -> Wai.Application
            app coreApp req respond =
-            case parseMethod $ Wai.requestMethod req of
-              Left _ ->
-                  respond invalidReq
-              Right stdMethod ->
-                  do let allActions = getMatchingRoutes stdMethod (Wai.pathInfo req)
-                     runResourceT $ withInternalState $ \st ->
-                         handleRequest stdMethod mLimit registryLift allActions st coreApp req respond
+            parseSpockMethod (Wai.requestMethod req) $ \method -> do
```
Comment (agrafix) [inline src/Web/Spock/Internal/Wire.hs]: Coding conventions of the project: 

```haskell
parseSpockmethod (Wai.requestMethod req) $
\method ->
do let allAction = -- ..
    -- ...
```

**Comment by agrafix** [comment]
Thanks, great patch! I've only found some small issues, before merging :-)

**Comment by lloucas-imvu** [comment]
Ill take care of this likely sometime tomorow

**Comment by agrafix** [comment]
Okay, thanks!

**Pull request merged**